### ISSUE TITLE
Images with alpha channel are incorrectly scaled

### ISSUE THREAD

**Issue by marcan** [obsproject/obs-studio#3055]
## Platform
Operating system and version: All (tested on Arch Linux and macOS)
OBS Studio version: 25.0.8

## Expected Behavior
Images with alpha channel should be correctly scaled, particularly along the transparency edges, without introducing extraneous colors.

## Current Behavior
OBS currently scales alpha channel images incorrectly, introducing dark or colored edges.

## Steps to Reproduce
1. Add a white color source
2. Add an Image source with [this image](https://user-images.githubusercontent.com/98387/84230471-f6721500-ab26-11ea-861f-9856454a848a.png).

Note that that image only has transparent and white pixels. Therefore, the resulting frame should be completely white, as all content is white on white.

3. Scale the image to any non-1:1 scale.

Result:

![result](https://user-images.githubusercontent.com/98387/84230558-2faa8500-ab27-11ea-8653-ca569eeaf09b.png)

## Additional information
This happens because OBS is currently taking the image with non-premultiplied alpha (as stored in the PNG), and scaling/sampling the color planes independently. This is mathematically incorrect: scaling non premultiplied alpha images is not separable.

The correct way to do this is to convert the image to premultiplied alpha, and then scale it (and composite it appropriately). This eliminates the artifacts and is mathematically correct.

Basically, if you ever have a texture with non-premultiplied alpha and you're sampling it at anything other than 1:1 scale with filtering, that's wrong. Any such instances in OBS should be replaced with a conversion to premultiplied alpha prior to sampling. [Here's](https://developer.nvidia.com/content/alpha-blending-pre-or-not-pre) some background on the issue.

**Comment by jpark37** [comment]
TLDR known issue, painful to fix. I fixed several bad cases last year, but this particular path is nasty to fix cleanly because it would require touching a lot of code.

The test PNG is actually ambiguous. Transparent black and opaque white are 0/0/0/0 and 1/1/1/1 regardless of straight or premultiplied. The color becomes unambiguous when bilinear filtering is applied by the GPU, interpolating to a premultiplied gray.

Rendering a scaled transparent source directly to the canvas is a known buggy code path where the blend state expects straight alpha from the pixel shader but receives premultiplied. You can work around this issue by setting Scale Filtering to anything other than Default. The clean fix would touch a lot of code, and OBS has more severe color issues in the backlog that should be addressed first, e.g. applying alpha to linear colors.

**Comment by marcan** [comment]
The test PNG is a standard non-premultiplied image (as required by the PNG spec), which means the colors are x/x/x/0 and 1/1/1/1 (in this case x=0, but it could be anything and the output shouldn't change). You are correct that premultiplying it would still result in 0/0/0/0 1/1/1/1, but the point is that then the *blending* should happen by interpreting it as premultiplied, and in that case the standard texture sampling would do the right thing.

Indeed turning on scale filtering fixes the issue. The problem is not alpha type confusion, though, as you can see if you try using [this image](https://user-images.githubusercontent.com/98387/84234731-ae0b2500-ab2f-11ea-9c6c-c6595aaa0796.png) which has a 50% transparent section. The bulk of the section is blended properly, which means the blend state is consistent with the alpha type. The only problem is the edges, so the bug must be only that the sampling is done on straight alpha, not that straight alpha is confused for premultiplied anywhere.

**Comment by jpark37** [comment]
However you want to frame the problem, the issue is known, and it won't be addressed before the more important ones are fixed.

**Comment by marcan** [comment]
Sure, I'm just filing it because I didn't see a similar bug open.

**Comment by sobotka** [comment]
> The test PNG is actually ambiguous. Transparent black and opaque white are 0/0/0/0 and 1/1/1/1 regardless of straight or premultiplied.

With associated (premultiplied) both occlusion and emission are expressed. Things like reflections, flares, glows, fires, etc. are entirely valid to be completely unoccluding, with alpha of zero, and emitting with nonzero RGB.

Common misconception with how alpha behaves.

**Comment by marcan** [comment]
Yup, that's one of the reasons why I think straight alpha is a bad idea (outside UIs). It doesn't let you express flare/bloom effects, while premultiplied does.

**Comment by jpark37** [comment]
I was just pointing out that even if I had performed a alpha multiplication on the original PNG data, it wouldn't fix the problem. (Obviously, since the data would be identical.) Doing the fix neatly with minimal hacks and without unnecessary draws is not trivial, and a workaround exists for the time being.

**Comment by alcomposer** [comment]
I have also experienced this issue for lower thirds alpha material.  prores4444 media (associated alpha) does not composite correctly:

https://github.com/obsproject/obs-studio/pull/1852#issuecomment-637359558

**Comment by marcan** [comment]
@alcomposer that's a different problem (premultiplied media being interpreted as straight alpha). The problem this issue is about is that the image is interpreted properly as straight alpha, and blended as such, but it is also naively *scaled* as straight alpha with simple GPU texture sampling, which is incorrect (GPU texture sampling is always incorrect for straight alpha, except for special cases where you can manually prepare the texture to work well anyway).

**Comment by PatTheMav** [comment]
FWIW I just tried the reproduction steps with OBS 28.0.3 on macOS today and wasn't able to reproduce - no matter how much I squished/squashed/scaled the image, it stayed practically invisible over the white colour source.

@gxalpha, @RytoEX feel free to confirm/correct.

**Comment by norihiro** [comment]
Unfortunately, I can reproduce the issue with the latest master, 2a76ffaf0.

OS: Fedora 35.
PNG file with alpha: [text-alpha.png](https://user-images.githubusercontent.com/780600/194589216-d55fe7b0-3689-4c6c-af7f-c399b6723409.png) I placed this image with 25.8 times scale.
Log file: https://obsproject.com/logs/MIXCAK6IH373F5VB
Screenshot of the preview:
![Screenshot 2022-10-08 00-16-55](https://user-images.githubusercontent.com/780600/194589552-5f01a3b5-c010-412d-94f6-6b132fee6f53.png)

**Comment by PatTheMav** [comment]
Right, with _that_ image I can reproduce it as well. Changing the source properties to "Apply alpha in linear space" seems to fix it at least visually, but maybe that's just a happy accident @jpark37?

**Comment by jamespark-unreal** [comment]
"Apply alpha in linear space" is the right way, but it's off by default for partial backwards compatibility. You can also get a nonlinear "web" blend without artifacts by leaving the linear checkbox unchecked, and using "Blending Method" -> "SRGB Off" which I didn't have the idea to add at the time. It's new to OBS 28.

Tldr, all of this behavior is expected and desired.

> Doing the fix neatly with minimal hacks and without unnecessary draws is not trivial, and a workaround exists for the time being.

I said this a while ago, but all of the color-correctness work around sRGB and HDR means that alpha can now be treated properly without hacks if you want to.

**Comment by jpark37** [comment]
"Apply alpha in linear space" is the right way, but it's off by default for partial backwards compatibility. You can also get a nonlinear "web" blend without artifacts by leaving the linear checkbox unchecked, and using "Blending Method" -> "SRGB Off" which I didn't have the idea to add at the time. It's new to OBS 28.

TLDR, all of this behavior is expected and desired.

> Doing the fix neatly with minimal hacks and without unnecessary draws is not trivial, and a workaround exists for the time being.

I said this a while ago, but all of the color-correctness work around sRGB and HDR means that alpha can now be treated properly without hacks if you want to.

**Comment by PatTheMav** [comment]
Cool so I guess the desired behaviour is achieved by choosing the right settings and we can close the issue.

**Comment by RytoEX** [comment]
Closing per above comments on intended workflows to reach desired outcomes.
